tensorflow/compiler/xla/service/gpu/fusion_merger.cc

```diff
@@ -214,15 +215,19 @@ Status FusionInstructionMerger::HandleFusion(HloInstruction* fusion) {
   if (fusion->fusion_kind() != HloInstruction::FusionKind::kLoop) {
     VLOG(3) << "Not merging " << fusion->name() << ": Is not loop fusion.";
     ++num_fail_not_loop_fusion_;
-    return Status::OK();
+    return false;
   }
 
   // Skip multiple output fusion. It's not yet supported.
   if (fusion->IsMultiOutputFusion()) {
     VLOG(3) << "Not merging " << fusion->name() << ": Is multi-output fusion.";
     ++num_fail_not_loop_fusion_;
-    return Status::OK();
+    return false;
   }
+}
+
+bool FusionInstructionMerger::IsProfitableToFuse(HloInstruction* fusion) {
+
   // Skip 'fusion' instruction if we cannot merge into all of its users.
```

Comment:
Couldn't we still fuse with duplication?

A->B
A->C

Say A can be fused into B but not C.
We could still do:
AB
A->C

"however, the result is so slow that we never want to do this"
Does that not make it a profitability criterion?

Reply:
> Couldn't we still fuse with duplication?
Yes that's possible. But mulit-output fusion is typically a better way to go. So you would do AB->C

> Does that not make it a profitability criterion?
You could argue it is one. You could also argue that the emitter is not designed for it and consider it unsupported.
Can you explain what the benefit of a clear distinction between legal / cost criteria will bring to the codebase? Is it relevant in the context of the ILP proposal you made?

Reply:
I think that in general separating legality and cost criteria is useful for experimentation.

For instance, it may be useful to enable all legal fusion to find the maximal fusion possible in a program.

Additionally, cost may be architecture specific and may change but legality conditions would likely remain constant. I'm not sure where the emitter support should fall.

For the ILP proposal, legality checks would be used to determine if an edge is contractible whereas the cost computations would be fed into the ILP solver instead.

Reply:
> For instance, it may be useful to enable all legal fusion to find the maximal fusion possible in a program.
> Additionally, cost may be architecture specific and may change but legality conditions would likely remain constant. I'm not sure where the emitter support should fall.

Whether fusions are legal is determined by whether emitters support them, i.e. we can lower them. This is platform-specific (CPU and GPU emitters differ) and has changed over time as we improved emitters.

> For the ILP proposal, legality checks would be used to determine if an edge is contractible whereas the cost computations would be fed into the ILP solver instead.

Some fusion decisions are based on heuristics that can be considered cost criteria, e.g. we are willing to read and write 10% more data to save the overhead of one kernel launch.

Reply:
I agree with both. Do you still prefer a single method over two separate ones for legality and cost?

Reply:
As I see it, merged_to_current_bytes_ratio is the only real cost-criteria here, i.e. the only thing you would want to tweak during experimentation. I am not convinced that extracting merged_to_current_bytes_ratio into a separate method is a big advantage, to be honest.